Improving `vmmap` output when debugging kernel

jxuanli opened the pull request:
## Improving `vmmap`

Demos are based off the challenge `The Jumps` from `NahamCon CTF 2025` (but with KASLR enabled for demo purposes.)

Currently, when running `qemu-system`, the `vmmap` command outputs:
![image](https://github.com/user-attachments/assets/a059d68e-d1f5-40a6-8c02-64fac11abaa2)

The `File` column contains only entries with the format `[pt_xxxxx...]` that are not useful. This PR improves the `vmmap` output when debug symbols are present.

## Demo

![image](https://github.com/user-attachments/assets/216903a5-21c2-4c91-9c86-0b8e89086e6d)

![image](https://github.com/user-attachments/assets/a5a8812c-8b10-40b2-945f-ee9e1dd61503)

jxuanli committed:
page walks

jxuanli committed:
Improving kernel-vmmap

jxuanli committed:
added test for kernel vmmap

jxuanli committed:
improve userland heap handling

jxuanli committed:
improve kernelland sections handling

jxuanli committed:
fixed typo

jxuanli committed:
fixed test

patryk4815 commented:
each page must have unique name, otherwise some commands may not work? 🤔 @disconnect3d do you remember what?

@jxuanli can you check with `set kernel-vmmap monitor`? I'm going to upstream some pt-dump changes into qemu - later pt-dump can be removed xD

jxuanli commented:
@patryk4815 its certainly trickier with the current `set kernel-vmmap monitor` (`monitor info mem`). i don't think it is useable yet without some very heavy coalescing on page gaps (or i guess we can omit printing out the `%esp stacks`, but that would not work if debug sysmbols are not present since I used kernel symbols to identify different regions). It also doesn't show if a page is executable -- one option is to find its PTE entry for every page but that's not very efficient and may not always work if debug symbols are not present. but I guess i should look into how `pt-dump` is implemented to perhaps get some ideas.

it might be more viable with your changes but we would then need to support older versions of qemu.

jxuanli committed:
adding support for info mem

disconnect3d commented:
Wow, this PR is awesome!

> each page must have unique name, otherwise some commands may not work? 🤔 @disconnect3d do you remember what?

This is about `search` command and few others (I think `probeleak` or `leakfind`?) where u want to target a specific page or pages. That's why we did `[pt_<part of address>]` initially.

Shall we maybe name those `userland{0,1,2,...}` or `userland_400`, `userland_401` etc for this reason?

disconnect3d commented on pwndbg/aglib/kernel/vmmap.py:
```diff
@@ -20,11 +20,131 @@
+        self.pages = pages
+        if not pwndbg.aglib.kernel.has_debug_syms():
+            return
+        # https://www.kernel.org/doc/Documentation/x86/x86_64/mm.txt
```
Please add a comment with kernel version (or versions)

jxuanli replied:
i believe this mapping has been pretty much the same for most recent kernels (v5.x, v6.x at least), but there might be outliers for different (non-x64) kernel builds, i ll add a comment once i figured that out

disconnect3d commented on pwndbg/aglib/vmmap.py:
```diff
@@ -34,3 +34,40 @@ def find(address: int | pwndbg.dbg_mod.Value | None) -> pwndbg.lib.memory.Page |
+
+    mappings = pages
+    for mapping in mappings:
+        # TODO: Check alignment
```
What about this?

jxuanli replied:
i think the mappings which are pages are by definition aligned

disconnect3d commented on tests/qemu-tests/tests/system/test_commands_kernel.py:
```diff
@@ -128,6 +128,22 @@ def test_command_msr_write():
+    res = gdb.execute("vmmap", to_string=True)
+    assert any(
+        key in res
+        for key in [
```
Use a tuple

jxuanli replied:
done

disconnect3d commented on pwndbg/color/memory.py:
```diff
@@ -88,10 +88,12 @@ def get(
         color = c.code
+    elif not page.write:
+        color = c.rodata
+    elif any(keyword in page.objfile for keyword in ["[heap", "physmap", "vmalloc"]):
```
```suggestion
    elif any(keyword in page.objfile for keyword in ("[heap", "physmap", "vmalloc")):
```

jxuanli replied:
done

disconnect3d commented on pwndbg/aglib/kernel/vmmap.py:
```diff
@@ -20,11 +20,131 @@
+        self.kbase = kbase = pwndbg.aglib.vmmap.find_kbase(pages)
+        if kbase is None:
+            return
+        self.sections = [
```
Can we use a tuple here?

jxuanli replied:
fixed

jxuanli committed:
changed array to tuple based on suggestions

patryk4815 commented:
> It also doesn't show if a page is executable

@jxuanli executable for x86-64 is fixed on my forked qemu https://github.com/patryk4815/qemu/commit/0621a6b9130ca21e1a89bbe3a2ab65a5e92756b8
`monitor info mem` right now is only supported in qemu for x86, x86-64 and riscv64

About `x` flag on upstream qemu: 
- riscv64 has `x` flags implemented (but our pwndbg code just omit it..)
- x86-32 - cpu don't support NX flag at all
- x86-64 - `x` flag is not implemented, I need send my patch https://github.com/patryk4815/qemu/commit/0621a6b9130ca21e1a89bbe3a2ab65a5e92756b8

jxuanli committed:
removing esp fixup stacks from display

jxuanli commented:
> This is about search command and few others (I think probeleak or leakfind?) where u want to target a specific page or pages. That's why we did [pt_<part of address>] initially.

how is userland handling this? iirc some of the userand pages have the same names (libc for example)

jxuanli commented:
@patryk4815 if your patch is merged in main branch then that prob works yea but we also need to support older versions of qemu?

patryk4815 commented:
@jxuanli yes, in pwndbg we will support older qemu version, qemu version can be detected by using `monitor info version`
btw do you see other bugs in `monitor info mem` that need be fixed?

jxuanli commented:
> btw do you see other bugs in monitor info mem that need be fixed?

~~no~~

jxuanli commented:
update: i have addressed some of the above comments. ~~In addition, I have decided to not include `%esp fixup stacks` in the outputs of `vmmap` because those are not useful in my experience  and will cluster up the display especially with `monitor info mem` (@patryk4815 can you try it, i don't think its necessarily a bug in qemu but rather `pt-dump`).~~ it turned out that i was wrong and it is used in some cases. i will have another PR soon on some more page helpers which will help the effort in supporting older versions of qemu when `kernel-vmmap` is set to `monitor`

jxuanli committed:
including call stacks

jxuanli commented:
btw when debugging a userland program, `vmmap` prints sections with the same name.
![image](https://github.com/user-attachments/assets/b45c691e-4053-4738-8dd3-d1b8ab10db9c)

jxuanli committed:
Merge branch 'dev' of https://github.com/pwndbg/pwndbg into kernel-vmmap

jxuanli committed:
Merge branch 'dev' of https://github.com/pwndbg/pwndbg into kernel-vmmap

jxuanli committed:
Merge branch 'dev' of https://github.com/pwndbg/pwndbg into pagehelpers

jxuanli committed:
implemented pagewalk

jxuanli committed:
added pagewalk test

jxuanli committed:
improved testing / output

jxuanli committed:
added docs

jxuanli committed:
renaming

jxuanli committed:
actually adding the remamed file

jxuanli committed:
Merge branch 'pagehelpers' into kernel-vmmap

jxuanli committed:
adding decoration for cpu arch

jxuanli committed:
Revert "adding decoration for cpu arch"

This reverts commit 84aa120f688d50a49868fde6de0b3d3276e23ce4.

jxuanli committed:
added arch check for pagewalk

jxuanli committed:
Merge branch 'pagehelpers' into kernel-vmmap

jxuanli committed:
adding req on symbols

jxuanli commented:
@patryk4815 per our prior discussion on the excessively many `%esp fixup stacks` being displayed by `mem info`, it appears that `pt-dump` was buggy. A pagewalk using the command from #3051 shows that those pages displayed from `mem info` are indeed present even though they are not shown by `pt-dump`.
![image](https://github.com/user-attachments/assets/baebb6a2-feb5-4d94-8b1d-45239867c3cf)

jxuanli committed:
supporting mem info

jxuanli committed:
refactored pagewalk helpers

jxuanli committed:
Merge branch 'pagehelpers' into kernel-vmmap

jxuanli committed:
added support for older versions of qemu-x86_64

jxuanli committed:
improved pagewalk helper function signature

jxuanli committed:
Merge branch 'pagehelpers' into kernel-vmmap

jxuanli committed:
improved processing of vmmap

jxuanli committed:
refactored a bit more

jxuanli committed:
refactored a bit more

jxuanli committed:
removed changes to kernel/vmmap

jxuanli committed:
Merge branch 'pagehelpers' into kernel-vmmap

jxuanli committed:
adding option to not process pages

jxuanli committed:
improving support to info mem

jxuanli committed:
changed to tuple

jxuanli committed:
changed to tuple

jxuanli committed:
changed to tuple

jxuanli committed:
Merge branch 'pagehelpers' into kernel-vmmap

jxuanli committed:
Merge branch 'pagehelpers' into kernel-vmmap

jxuanli committed:
Merge branch 'pagehelpers' into kernel-vmmap

jxuanli commented:
this branch is somewhat ready, though it depends on #3051.

it handles both modes of `kernel-vmmap` although we need to discuss more about how we should handle the excessive amount of entries printed by `info mem`. The handling of `info mem` could also benefit from #3065 if we decide to omit most of the `%esp fixup stacks`.

disconnect3d commented:
@jxuanli pls fix conflicts

jxuanli committed:
Merge branch 'dev' of https://github.com/pwndbg/pwndbg into kernel-vmmap

jxuanli committed:
added aarch64 mem mapping

jxuanli committed:
improved testing

jxuanli committed:
fixing for arm64

jxuanli committed:
a bit more amending

jxuanli commented:
@disconnect3d fixed and also did some fixing so that the output for aarch64 is sane  (updated the test to assert some of those changes)

jxuanli committed:
fixing test
